File: packages/functions/ios/Plugin/FirebaseFunctionsPlugin.swift
import Foundation
import Capacitor

/**
 * Please read the Capacitor iOS Plugin Development Guide
 * here: https://capacitorjs.com/docs/plugins/ios
 */
@objc(FirebaseFunctionsPlugin)
public class FirebaseFunctionsPlugin: CAPPlugin {
    public let tag = "FirebaseFunctions"
    public let errorNameMissing = "name must be provided."
    public let errorHostMissing = "host must be provided."
    private var implementation: FirebaseFunctions?

    override public func load() {
        self.implementation = FirebaseFunctions(plugin: self)
    }

    @objc func callByName(_ call: CAPPluginCall) {
        guard let name = call.getString("name") else {
            call.reject(errorNameMissing)
            return
        }
        let region = call.getString("region")
        let data = call.getValue("data")

        let options = CallByNameOptions(name: name, region: region, data: data)

        implementation?.callByName(options, completion: { result, error in
            if let error = error {
                CAPLog.print("[", self.tag, "] ", error)
                call.reject(error.localizedDescription)
                return
            }
            if let result = result?.toJSObject() as? JSObject {
                call.resolve(result)
            }
        })
    }

    @objc func callByUrl(_ call: CAPPluginCall) {
        guard let url = call.getString("url") else {
            call.reject("url must be provided.")
            return
        }
        let data = call.getValue("data")

        let options = CallByUrlOptions(url: url, data: data)

        implementation?.callByUrl(options, completion: { result, error in
            if let error = error {
                CAPLog.print("[", self.tag, "] ", error)
                call.reject(error.localizedDescription)
                return
            }
            if let result = result?.toJSObject() as? JSObject {
                call.resolve(result)
            }
        })
    }

    @objc func useEmulator(_ call: CAPPluginCall) {
        guard let host = call.getString("host") else {
            call.reject(errorHostMissing)
            return
        }
        let port = call.getInt("port") ?? 5001

        implementation?.useEmulator(host, port)
        call.resolve()
    }
}

File: packages/functions/android/src/main/java/io/capawesome/capacitorjs/plugins/firebase/functions/FirebaseFunctionsPlugin.java
package io.capawesome.capacitorjs.plugins.firebase.functions;

import com.getcapacitor.JSObject;
import com.getcapacitor.Logger;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;
import io.capawesome.capacitorjs.plugins.firebase.functions.classes.options.CallByNameOptions;
import io.capawesome.capacitorjs.plugins.firebase.functions.classes.options.CallByUrlOptions;
import io.capawesome.capacitorjs.plugins.firebase.functions.interfaces.NonEmptyResultCallback;
import io.capawesome.capacitorjs.plugins.firebase.functions.interfaces.Result;

@CapacitorPlugin(name = "FirebaseFunctions")
public class FirebaseFunctionsPlugin extends Plugin {

    public static final String TAG = "FirebaseFunctions";
    public static final String ERROR_NAME_MISSING = "name must be provided.";
    public static final String ERROR_URL_MISSING = "url must be provided.";

    private FirebaseFunctions implementation;

    public void load() {
        implementation = new FirebaseFunctions(this);
    }

    @PluginMethod
    public void callByName(PluginCall call) {
        try {
            JSObject json = call.getData();
            String name = json.getString("name", null);
            if (name == null) {
                call.reject(ERROR_NAME_MISSING);
                return;
            }
            String region = json.getString("region", null);
            Object data = json.opt("data");

            CallByNameOptions options = new CallByNameOptions(name, region, data);
            NonEmptyResultCallback callback = new NonEmptyResultCallback() {
                @Override
                public void success(Result result) {
                    call.resolve(result.toJSObject());
                }

                @Override
                public void error(Exception exception) {
                    Logger.error(TAG, exception.getMessage(), exception);
                    call.reject(exception.getMessage());
                }
            };

            implementation.callByName(options, callback);
        } catch (Exception exception) {
            Logger.error(TAG, exception.getMessage(), exception);
            call.reject(exception.getMessage());
        }
    }

    @PluginMethod
    public void callByUrl(PluginCall call) {
        try {
            JSObject json = call.getData();
            String url = json.getString("url");
            if (url == null) {
                call.reject(ERROR_URL_MISSING);
                return;
            }
            Object data = json.opt("data");

            CallByUrlOptions options = new CallByUrlOptions(url, data);
            NonEmptyResultCallback callback = new NonEmptyResultCallback() {
                @Override
                public void success(Result result) {
                    call.resolve(result.toJSObject());
                }

                @Override
                public void error(Exception exception) {
                    Logger.error(TAG, exception.getMessage(), exception);
                    call.reject(exception.getMessage());
                }
            };

            implementation.callByUrl(options, callback);
        } catch (Exception exception) {
            Logger.error(TAG, exception.getMessage(), exception);
            call.reject(exception.getMessage());
        }
    }
}
